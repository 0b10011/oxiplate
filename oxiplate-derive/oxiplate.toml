# The macro by itself (`oxiplate_derive`) doesn't support escaping.
# This is only for tests directed at the macro.
# The main library (`oxiplate`) supported and tests escaping.
# Libraries and binaries using Oxiplate 
# will usually rely on inferring the escaper from file extensions,
# but they may also specify their own fallback escaper,
# possibly even "raw" for plain text use cases.
fallback_escaper_group = "raw"

# The macro by itself doesn't support escaping,
# and due to feature unification,
# when `cargo test` is run on the workspace,
# the `oxiplate` feature is turned on
# resulting in inferrence happening.
infer_escaper_group_from_file_extension = false
