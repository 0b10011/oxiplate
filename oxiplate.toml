default_escaper_group = "html"

escaper_groups.html.escaper = "::oxiplate::escapers::html::HtmlEscaper"
escaper_groups.md.escaper = "::oxiplate::escapers::markdown::MarkdownEscaper"
