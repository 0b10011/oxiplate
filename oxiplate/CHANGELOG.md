# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.3](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.2.2...oxiplate-v0.2.3) - 2025-01-25

### Fixed

- more accurate spans on parent templates

### Other

- support for subdirectories in oxiplate broken tests
- clippy fixes

## [0.2.2](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.2.1...oxiplate-v0.2.2) - 2025-01-14

### Fixed

- extends now uses `oxiplate` instead of `oxiplate_derive` to avoid an extra import by the user

## [0.2.1](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.2.0...oxiplate-v0.2.1) - 2025-01-13

### Fixed

- raw escaping when default escaper is set and escaping non-string slices

## [0.2.0](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.12...oxiplate-v0.2.0) - 2025-01-13

### Added

- [**breaking**] prevent whitespace replace characters when there's no matching whitespace to replace

### Other

- simplify whitespace adjustment checks in end tags

## [0.1.12](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.11...oxiplate-v0.1.12) - 2025-01-13

### Added

- fail to compile if adjacent tags with only whitespace between explicitly specify different whitespace adjustment characters
- only pass braces in as arguments instead of entire strings that contain them

### Other

- add more text to the format injection text to better show changes with normal text
- reduce item count for static text with whitespace

## [0.1.11](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.10...oxiplate-v0.1.11) - 2025-01-10

### Added

- include the parent contents in an overridden block
- support for `{% else %}` in for loops

### Other

- update Cargo.lock dependencies
- remove `Cargo.lock` for the macro and ensure another doesn't get committed
- conditionally compile expansion tests to make it clear they will fail on old rustc versions

## [0.1.10](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.9...oxiplate-v0.1.10) - 2025-01-09

### Added

- combine static strings in concats into the format string for smaller templates
- concat operator (~)

### Fixed

- string continuation escapes and spans after escape sequences

### Other

- merge variable tests into a single file
- clear `/actual/` each run, error for leftover files, and fix the check against `expansion.rs`
- turn on `format_strings` and reformat
- updated expected expansion after rust update

## [0.1.9](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.8...oxiplate-v0.1.9) - 2025-01-08

### Fixed

- building templates when `std` is overridden

### Other

- add additional rustfmt rules and reformat
- synchronize versions
- cleaned up the attribute/struct parsing code even more
- split out state, add some documentation, and clean up parsing code a bit
- update expected calc expansion
- cover more complicated calculations
- cover more complicated cases of `||` and `&&` checks
- add docs for the derive macro
- moved `Source` and `SourceOwned` into its own module

## [0.1.8](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.6...oxiplate-v0.1.8) - 2025-01-06

Merged the macro back in from https://github.com/0b10011/oxiplate-derive/commit/7c965e47dc0714a2e10b9bd8d4f488c101224fff

### Added

- added support for rust keywords as identifiers and cleaned up code along the way
- improved error message for missing `if` expression
- shortened source range for errors that don't provide a span
- building the write format with the templates themselves to reduce the number of arguments needed
- calling `write_str()` instead of `write_fmt()` for a single static token
- combined sequential static text and whitespace into a single concat
- combined sequential static text, whitespace, and writs into a single write call

### Fixed

- improved span information for field/method access
- use correct module for `escape()`
- pass escaper by reference to match expectation in the main crate

### Other

- try using package-specific versions instead
- bring back separate changelog for the macro and fix versions
- move the derive macro back to the main repo
- rewrite readme to add badges and point to book/docs
- `cargo clippy` changes
- Fix newlines
- Move templates to their own directory and support nesting
- Fix hygiene
- Initial attempt to use `include_str!()` to fetch files
- Add message about code being experimental
- Add support for templates stored in separate files
- Add basic documentation
- remove some unused code
- prevent clippy test output from changing based on whether build is required
- check if spans work with clippy
- dependency updates
- fixed repository link
- changed expansion tests to fail when the expected output for a test is missing
- build the path to `oxiplate.toml` from the env instead to help with testing


## [0.1.6](https://github.com/0b10011/oxiplate/compare/oxiplate-v0.1.5...oxiplate-v0.1.6) - 2025-01-01

### Added

- add an escaper for markdown

### Other

- extend `cargo dev` to build docs as well
- fix path in example
- add more tests and refactor a bit

## [0.1.5](https://github.com/0b10011/oxiplate/compare/v0.1.4...v0.1.5) - 2024-12-31

### Added

- add support for custom escapers and move macro to `oxiplate-derive` crate

### Fixed

- use correct name for escaper group in message

### Other

- use an enum for escapers to reduce boilerplate

## [0.1.4](https://github.com/0b10011/oxiplate/compare/v0.1.3...v0.1.4) - 2024-12-30

### Other

- add basic support for custom escapers
- add state to escaper and rename it to prep for external escapers
- add state and move local variables into it

## [0.1.3](https://github.com/0b10011/oxiplate/compare/v0.1.2...v0.1.3) - 2024-12-25

### Other

- add writs doc
- add "Getting started" doc

## [0.1.2](https://github.com/0b10011/oxiplate/compare/v0.1.1...v0.1.2) - 2024-12-23

### Other

- add start of book

## [0.1.1](https://github.com/0b10011/oxiplate/compare/v0.1.0...v0.1.1) - 2024-12-23

### Other

- add funding info
- add release-plz for release management
