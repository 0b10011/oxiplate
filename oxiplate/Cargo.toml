[package]
name = "oxiplate"
description.workspace = true
version.workspace = true
authors.workspace = true
homepage.workspace = true
readme.workspace = true
repository.workspace = true
categories.workspace = true
keywords.workspace = true
license.workspace = true
edition.workspace = true

[lints]
workspace = true

[dependencies]
itoa = "1.0.15"
oxiplate-derive = { version = "0.8.1", path = "../oxiplate-derive", default-features = false, features = ["oxiplate"] }

[dev-dependencies]
rustversion = "1.0.19"
trybuild = "1.0.101"

[features]
default = ["built-in-escapers", "config"]

# Automatically adds the built-in escapers to the list of available escapers,
# and uses "html" as the default escaper if a different one isn't specified.
built-in-escapers = ["oxiplate-derive/built-in-escapers"]

# Makes it possible to configure Oxiplate via `/oxiplate.toml`.
config = ["oxiplate-derive/config"]

# Feature only meant to be turned on for tests.
# Changes the output for various types to 
debug-fast-escape-type-priority = ["oxiplate-derive/debug-fast-escape-type-priority"]
