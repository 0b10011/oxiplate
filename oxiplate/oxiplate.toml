escaper_groups.your_group.escaper = "::oxiplate::escapers::your_group::YourEscaper"