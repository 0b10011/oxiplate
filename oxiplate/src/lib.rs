#![doc(issue_tracker_base_url = "https://github.com/0b10011/oxiplate/issues/")]
#![doc(test(no_crate_inject))]
#![doc(test(attr(deny(warnings))))]
#![doc = include_str!("../README.md")]
#![warn(missing_docs)]

use std::io::{Error, Write};

pub use oxiplate_derive::Oxiplate;

pub mod escapers;

/// Optimized render function trait.
pub trait Render {
    /// Estimated output length of the template.
    const ESTIMATED_LENGTH: usize;

    /// Render the template into a string.
    ///
    /// # Errors
    ///
    /// If strings cannot be written to the formatter.
    fn render(&self) -> Result<String, Error> {
        let mut string = Vec::with_capacity(Self::ESTIMATED_LENGTH);
        self.render_into(&mut string)?;
        unsafe {
            Ok(String::from_utf8_unchecked(string))
        }
    }

    /// Render the template into a writer.
    ///
    /// # Errors
    ///
    /// If strings cannot be written to the formatter.
    fn render_into<W: Write>(&self, writer: &mut W) -> ::std::io::Result<()>;
}
