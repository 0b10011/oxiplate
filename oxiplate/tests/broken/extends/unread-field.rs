use oxiplate::{Oxiplate, Render};

#[deny(dead_code)]
#[derive(Oxiplate)]
#[oxiplate = "./multiple-blocks-inner.html.oxip"]
struct Data {
    unread_field: String,
}

fn main() {
    let data = Data {
        unread_field: String::from(
            "this isn't read in the template, but will be included in the parent template's \
             extends data and should still cause an error",
        ),
    };

    assert_eq!(
        data.render().unwrap(),
        "<!DOCTYPE html>\n<header>header</header>\n<main>main</main>\\
         n<footer>footer</footer>"
    );
}
