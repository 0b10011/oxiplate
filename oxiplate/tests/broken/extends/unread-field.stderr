error: field `unread_field` is never read
 --> tests/broken/extends/unread-field.rs:7:5
  |
6 | struct Data {
  |        ---- field in this struct
7 |     unread_field: String,
  |     ^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> tests/broken/extends/unread-field.rs:3:8
  |
3 | #[deny(dead_code)]
  |        ^^^^^^^^^
